### PR TITLE
Generate debug files (as read-only VS Code views)

### LINKED ISSUE #307: Generate debug files (as read-only VS Code views)

### PR DESCRIPTION
**Pull request by JonasWanke**
Closes: #307

Only HIR, AST, and RCST are implemented at all, and these are also just initial implementations. Features like folding, highlights/references and go to definition are still missing.

RCST:

image

AST:

image

HIR:

image

### Checklist

- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

### CODE REVIEW

File: compiler/frontend/src/rich_ir.rs [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+}
+
+#[derive(Debug, Hash, PartialEq, Eq, Clone, Copy)]
+pub enum TokenType {
+    Int,
+    Symbol,
+    Text,
+}
+#[derive(Debug, EnumSetType)]
+pub enum TokenModifier {}
```
Comment (MarcelGarus) [inline compiler/frontend/src/rich_ir.rs]: What's this for?

File: compiler/frontend/src/ast.rs [hunk 2]
```diff
@@ -392,181 +383,174 @@ impl CollectErrors for Vec<Ast> {
...
-                .iter()
-                .map(|it| format!("{it}"))
-                .join("\n")
-                .lines()
-                .map(|line| format!("  {line}"))
-                .join("\n")
+impl ToRichIr for Int {
+    fn build_rich_ir(&self, builder: &mut RichIrBuilder) {
+        builder.push(
+            format!(" int {}", self.0),
```
Comment (MarcelGarus) [inline compiler/frontend/src/ast.rs]:
```suggestion
            format!("int {}", self.0),
```

File: compiler/language_server/src/features_candy/mod.rs [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+    diagnostics_sender: Sender<(Module, Vec<Diagnostic>)>,
+    hints_events_sender: Sender<hints::Event>,
+}
+impl CandyFeatures {
+    pub fn new(
+        diagnostics_sender: Sender<(Module, Vec<Diagnostic>)>,
+        hints_sender: Sender<(Module, Vec<Hint>)>,
+    ) -> Self {
+        let (hints_events_sender, hints_events_receiver) = tokio::sync::mpsc::channel(1024);
+        // let (hints_sender, mut hints_receiver) = tokio::sync::mpsc::channel(1024);
```
Comment (MarcelGarus) [inline compiler/language_server/src/features_candy/mod.rs]:
```suggestion
```

File: compiler/language_server/src/features_candy/mod.rs [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+    }
+    fn semantic_tokens<'life0, 'life1, 'async_trait>(
+        &'life0 self,
+        db: &'life1 Database,
+        module: Module,
+    ) -> Pin<Box<dyn Future<Output = Vec<SemanticToken>> + Send + 'async_trait>>
+    where
+        'life0: 'async_trait,
+        'life1: 'async_trait,
+        Self: 'async_trait,
```
Comment (MarcelGarus) [inline compiler/language_server/src/features_candy/mod.rs:193]:
```suggestion
    fn semantic_tokens<'self, 'db, 'async_trait>(
        &'self self,
        db: &'db Database,
        module: Module,
    ) -> Pin<Box<dyn Future<Output = Vec<SemanticToken>> + Send + 'async_trait>>
    where
        'self: 'async_trait,
        'db: 'async_trait,
        Self: 'async_trait,
```

File: compiler/language_server/src/server.rs [hunk 5]
```diff
@@ -0,0 +1,458 @@
...
+        let (hints_sender, mut hints_receiver) = tokio::sync::mpsc::channel(1024);
+
+        let (service, client) = LspService::build(|client| {
+            let candy_features = CandyFeatures::new(diagnostics_sender, hints_sender);
+            let ir_features = Ir::iter().map(|ir| (ir, IrFeatures::new(ir))).collect();
+
+            // let (mut diagnostics_receiver, mut hints_receiver) = {
+            //     let mut receivers = self.diagnostics_and_hints_receiver.lock().await;
+            //     mem::take(&mut *receivers).unwrap()
+            // };
```
Comment (MarcelGarus) [inline compiler/language_server/src/server.rs]: Is this still needed?

**Review by MarcelGarus — CHANGES_REQUESTED** [review]

File: compiler/language_server/src/features_candy/mod.rs (hunk 4 above)
Comment (JonasWanke) [reply compiler/language_server/src/features_candy/mod.rs:193]: `self` is a reserved name and can't be used as a lifetime name. Also, the names have to be the same as in the trait itself, so we can't change them here without also having to manually define the async function there

**Review by MarcelGarus — APPROVED** [review]

**Pull request merged**